src/main/java/jenkins/plugins/slack/SlackUserIdResolver.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * The MIT License
+ *
+ * Copyright (C) 2019 Electronic Arts Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.plugins.slack;
+
+import hudson.model.AbstractBuild;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+import hudson.tasks.MailAddressResolver;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import jenkins.scm.RunWithSCM;
+import org.apache.commons.lang.StringUtils;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpStatus;
+import org.apache.http.ParseException;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class SlackUserIdResolver {
+
+    private static final Logger LOGGER = Logger.getLogger(SlackUserIdResolver.class.getName());
+
+    private static final String LOOKUP_BY_EMAIL_METHOD_URL_FORMAT = "https://slack.com/api/users.lookupByEmail?token=%s&email=%s";
```

Comment:
This is going down a slippery slope IMHO :sweat: 
I'd much prefer if users or a script could add slack ids to their Jenkins user profile.
That way you also give them an option to opt-out by removing their slack id.

Reply:
@casz Would the following alternative / compromise be acceptable?
- Create a new `SlackUserProperty` that caches the slack user id.
- If the property exists for the current commit author, use the associated slack user id
- If the property doesn't exist for the current commit author, create a new one and use the `lookUpByEmail` API method to retrieve the slack user id
- If a user wants to opt-out, they can clear the slack id from their user profile

Reply:
It works for me, I'm not hugely concerned about opt-out, its company internal and they probably should be getting these notifications, if they don't want them at all then they can make their github and slack emails not match (possibly).

but would be good to allow an opt-out if it's not too difficult and doesn't make the common case hard

Reply:
Seems reasonable with current SlackUserProperty 👍